editor/header/tools/publish-button.js

```diff
@@ -0,0 +1,125 @@
+/**
+ * External dependencies
+ */
+import { connect } from 'react-redux';
+import { get } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import Button from 'components/button';
+
+function PublishButton( {
+	blocks,
+	post,
+	isSuccessful,
+	isRequesting,
+	isError,
+	requestIsNewPost,
+	onUpdate,
+	onPublish
+} ) {
+	let buttonEnabled = true;
+	let buttonText, saveCallback;
+
+	if ( isRequesting ) {
+		buttonEnabled = false;
+		if ( requestIsNewPost ) {
+			buttonText = wp.i18n.__( 'Publishing...' );
+		} else {
+			buttonText = wp.i18n.__( 'Updating...' );
+		}
+	} else if ( isSuccessful ) {
+		if ( requestIsNewPost ) {
+			buttonText = wp.i18n.__( 'Published!' );
+		} else {
+			buttonText = wp.i18n.__( 'Updated!' );
+		}
+	} else if ( isError ) {
+		if ( requestIsNewPost ) {
+			buttonText = wp.i18n.__( 'Publish failed' );
+		} else {
+			buttonText = wp.i18n.__( 'Update failed' );
+		}
+	} else if ( post && post.id ) {
+		buttonText = wp.i18n.__( 'Update' );
+	} else {
+		buttonText = wp.i18n.__( 'Publish' );
+	}
```

Comment:
I'd probably change this to a single `if` chain without nesting instead, like `if ( isRequesting && requestIsNewPost )`.  I'll try it